sys/arm64/vmm/vmm.c

```diff
@@ -0,0 +1,1818 @@
+{
+	int error;
+
+	if (reg >= VM_REG_LAST)
+		return (EINVAL);
+	error = vmmops_setreg(vcpu->cookie, reg, val);
+	if (error || reg != VM_REG_GUEST_PC)
+		return (error);
+
+#if __has_feature(capabilities)
+#ifdef __CHERI_PURE_CAPABILITY__
+	vcpu->nextpc = (uintcap_t)cheri_setaddress(kernel_root_cap, val);
+#else
+	vcpu->nextpc = (uintcap_t)cheri_setaddress(cheri_getpcc(), val);
+#endif
+#else
+	vcpu->nextpc = val;
+#endif
+
+	return(0);
+}
+
+void *
+vm_get_cookie(struct vm *vm)
+{
+	return vm->cookie;
+}
+
+int
+vm_inject_exception(struct vcpu *vcpu, uint64_t esr, uint64_t far)
+{
+	return (vmmops_exception(vcpu->cookie, esr, far));
+}
+
+int
+vm_attach_vgic(struct vm *vm, struct vm_vgic_descr *descr)
+{
+	return (vgic_attach_to_vm(vm->cookie, descr));
+}
+
+int
+vm_assert_irq(struct vm *vm, uint32_t irq)
+{
+	return (vgic_inject_irq(vm->cookie, -1, irq, true));
+}
+
+int
+vm_deassert_irq(struct vm *vm, uint32_t irq)
+{
+	struct hyp *hyp = (struct hyp *)vm->cookie;
+	int error;
+
+	error = vgic_inject_irq(hyp, -1, irq, false);
+
+	return (error);
+}
+
+int
+vm_raise_msi(struct vm *vm, uint64_t msg, uint64_t addr, int bus, int slot,
+    int func)
+{
+	/* TODO: Should we raise an SError? */
+	return (vgic_inject_msi(vm->cookie, msg, addr));
+}
+
+static int
+vm_handle_smccc_call(struct vcpu *vcpu, struct vm_exit *vme, bool *retu)
+{
+	struct hypctx *hypctx;
+	int i;
+
+	hypctx = vcpu_get_cookie(vcpu);
+
+	if ((hypctx->tf.tf_esr & ESR_ELx_ISS_MASK) != 0)
+		return (1);
+
+	vme->exitcode = VM_EXITCODE_SMCCC;
+	vme->u.smccc_call.func_id = hypctx->tf.tf_x[0];
+	for (i = 0; i < nitems(vme->u.smccc_call.args); i++)
+		vme->u.smccc_call.args[i] = hypctx->tf.tf_x[i + 1];
+
+	*retu = true;
+	return (0);
+}
+
+static int
+vm_handle_wfi(struct vcpu *vcpu, struct vm_exit *vme, bool *retu)
+{
+	vcpu_lock(vcpu);
+	while (1) {
+		if (vgic_has_pending_irq(vcpu->cookie))
+			break;
+
+		if (vcpu_should_yield(vcpu))
+			break;
+
+		vcpu_require_state_locked(vcpu, VCPU_SLEEPING);
+		/*
+		 * XXX msleep_spin() cannot be interrupted by signals so
+		 * wake up periodically to check pending signals.
+		 */
+		msleep_spin(vcpu, &vcpu->mtx, "vmidle", hz);
+		vcpu_require_state_locked(vcpu, VCPU_FROZEN);
+	}
+	vcpu_unlock(vcpu);
+
+	*retu = false;
+	return (0);
+}
+
+static int
+vm_handle_paging(struct vcpu *vcpu, bool *retu)
+{
+	struct vm *vm = vcpu->vm;
+	struct vm_exit *vme;
+	struct vm_map *map;
+	uint64_t addr, esr;
+	pmap_t pmap;
+	int ftype, rv;
+
+	vme = &vcpu->exitinfo;
+
+	pmap = vmspace_pmap(vcpu->vm->vmspace);
+	addr = vme->u.paging.gpa;
+	esr = vme->u.paging.esr;
+
+	/* The page exists, but the page table needs to be upddated */
+	if (pmap_fault(pmap, esr, addr) == KERN_SUCCESS)
+		return (0);
+
+	switch (ESR_ELx_EXCEPTION(esr)) {
+	case EXCP_INSN_ABORT_L:
+	case EXCP_DATA_ABORT_L:
+		ftype = VM_PROT_EXECUTE | VM_PROT_READ | VM_PROT_WRITE;
+#if __has_feature(capabilities)
+		ftype |= VM_PROT_READ_CAP | VM_PROT_WRITE_CAP;
```

Comment:
The pmap/vmspace in question belongs to the VM itself, not a user process, so I don't think revocation enters the picture, much as it doesn't (or at least shouldn't) for `kernel_pmap`.

Reply:
We still don't pass in READ_CAP/WRITE_CAP for vm_fault_trap when talking about kernel mappings though, so I wouldn't like to say it's a good idea to unilaterally pass it in when it's not a capability fault...

Reply:
Yes, but the kernel has fine-grained VM map entries which limit protections on a given VA range; there is no such thing here for the GPA space which is getting mapped here. The only thing to do were a capability data abort to arise on a valid GPA would be to upgrade protections on the stage 2 entry, so preemptively setting all of these prot bits doesn't seem like an outrageous thing to do even though it's certainly questionable.

Is there a specific reason we should be more particular here? I can implement finer-grained handling of exceptions of course but I want to be sure I understand what goes wrong if we don't.

Reply:
I don't think so, so long as you never hit this path for an object that doesn't have OBJ_HASCAP, since VM_OBJECT_ASSERT_CAP will trip up various paths if so. Would rather see this and the upstream code become more precise, though, as it's a bit hacky.

Reply:
We set OBJ_HASCAP on the VM objects that back the guest physical address space.

I will add a comment about this in the upstream version.